eng/MultiTargetRoslynComponent.targets.template

```diff
@@ -0,0 +1,41 @@
+<Project>
+  <!-- Reads the version of the compiler APIs that are currently being used in order to pick the correct Roslyn components. -->
+  <Target Name="_{TargetPrefix}ResolveRoslynApiMajorVersion"
+          Condition="'$(_RoslynApiMajorVersion)' == ''">
+    <GetAssemblyIdentity AssemblyFiles="$(RoslynTargetsPath)\Microsoft.Build.Tasks.CodeAnalysis.dll">
+      <Output TaskParameter="Assemblies" ItemName="_CodeAnalysisIdentity" />
+    </GetAssemblyIdentity>
+
+    <PropertyGroup>
+      <_RoslynApiMajorVersion>$([System.Version]::Parse(%(_CodeAnalysisIdentity.Version)).Major)</_RoslynApiMajorVersion>
+    </PropertyGroup>
+  </Target>
+
+  <Target Name="_{TargetPrefix}RoslynComponentMultiTargeting" 
+          Condition="'$(SupportsRoslynComponentVersioning)' != 'true'" 
+          AfterTargets="ResolveAssemblyReferences" 
```

Comment:
This might be too late.  Wouldn't conflict resolution run before this and drop one of them?  I guess that would depend on if we gave them the same name or not.  Actually maybe we need to give different names per @sharwell's sxs versioning suggestion.

Reply:
In my testing, it isn't too late. `_HandlePackageFileConflicts` outputs:

`Encountered conflict between 'Analyzer:C:\Users\eerhardt\.nuget\packages\system.text.json\7.0.0-dev\analyzers\dotnet\roslyn3.9\cs\System.Text.Json.SourceGeneration.dll' and 'Analyzer:C:\Users\eerhardt\.nuget\packages\system.text.json\7.0.0-dev\analyzers\dotnet\roslyn4.0\cs\System.Text.Json.SourceGeneration.dll'. Could not determine winner due to equal file and assembly versions.`

So both Analyzers are still around during `ResolveAssemblyReferences`.

However, I can move this up to be `AfterTargets="ResolvePackageDependenciesForBuild"`. Or potentially even `ResolveLockFileAnalyzers`. Thoughts?

Reply:
We need your logic to run before conflict resolution.  The fact that you are seeing `Encountered conflict ` means it's not early enough.  Your logic needs to behave "as if" it were NuGet making the decision.  `AfterTargets` the target(s) that raise(s) these analyzers from NuGet packages would be best if it is / they are public.

Reply:
Both `ResolvePackageDependenciesForBuild` and `ResolveLockFileAnalyzers` happen before conflict resolution runs. And `ResolvePackageDependenciesForBuild` depends on `ResolveLockFileAnalyzers`.

The flow, as I understand it is:

1. [ResolvePackageAssets](https://github.com/dotnet/sdk/blob/17181d34b143cbf4b957abdaabb1d2fb116dc612/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets#L246-L275) outputs the analyzers in the project.assets.json to the `@(ResolvedAnalyzers)` item.
2. [ResolveLockFileAnalyzers](https://github.com/dotnet/sdk/blob/17181d34b143cbf4b957abdaabb1d2fb116dc612/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets#L485-L490) depends on `ResolvePackageAssets` and turns `@(ResolvedAnalyzers)` into `@(Analyzer)`
3. [ResolvePackageDependenciesForBuild](https://github.com/dotnet/sdk/blob/17181d34b143cbf4b957abdaabb1d2fb116dc612/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets#L135-L149) depends on `ResolveLockFileAnalyzers`, and doesn't have the "LockFile" in the name.

Another thing occurred to me as I wrote this - none of these targets occur with a non-SDK .csproj that uses `PackageReference`. Do you think that is a scenario we need to support?

Reply:
> none of these targets occur with a non-SDK .csproj that uses PackageReference. Do you think that is a scenario we need to support?

I believe that scenario would still raise analyzers and run in VS 2022, so yes.  Our packages support install in classic .NETFramework projects.  The nice thing about `AfterTargets` is that it's a noop if the target doesn't exist, and it'll only run once for a given project/global-property-set.

Reply:
Thanks for the help. I changed this target to run `AfterTargets="ResolvePackageDependenciesForBuild;ResolveNuGetPackageAssets"`, which should address these concerns.